src/sentry/api/endpoints/organization_events_trace.py

```diff
@@ -17,20 +20,83 @@
 MAX_TRACE_SIZE = 100
 NODESTORE_KEYS = ["timestamp", "start_timestamp"]
 
+SnubaEvent = Mapping[str, str]
+TraceError = Mapping[str, str]
+ResponseType = Dict[str, Any]
+
+
+class TraceEvent:
+    def __init__(self, event: SnubaEvent, parent: Optional[str], generation: Optional[int]) -> None:
+        self.event: SnubaEvent = event
+        self.parent_event_id: Optional[str] = parent
+        # Can be None on the light trace when we don't know the parent
+        self.generation: Optional[int] = generation
+        self.nodestore_event: Optional[Event] = None
+        self.errors: List[TraceError] = []
+        self.children: List[TraceEvent] = []
+
+    def to_dict(self, full: bool = False, detailed: bool = False) -> ResponseType:
+        result: ResponseType = {
+            "event_id": self.event["id"],
+            "span_id": self.event["trace.span"],
+            "transaction": self.event["transaction"],
+            "transaction.duration": self.event["transaction.duration"],
+            "transaction.op": self.event["transaction.op"],
+            "project_id": self.event["project.id"],
+            "project_slug": self.event["project"],
+            # Avoid empty string for root self.events
+            "parent_span_id": self.event["trace.parent_span"] or None,
+            "parent_event_id": self.parent_event_id,
+            "generation": self.generation,
+            "errors": self.errors,
+        }
+        if full:
+            if detailed and "transaction.status" in self.event:
+                result.update(
+                    {
+                        "transaction.status": SPAN_STATUS_CODE_TO_NAME.get(
+                            self.event["transaction.status"], "unknown"
+                        ),
+                    }
+                )
+            if self.nodestore_event:
+                result.update(
+                    {
+                        event_key: self.nodestore_event.data.get(event_key)
+                        for event_key in NODESTORE_KEYS
+                    }
+                )
+                if detailed:
+                    if "measurements" in self.nodestore_event.data:
+                        result["measurements"] = self.nodestore_event.data.get("measurements")
+                    result["_meta"] = {}
+                    result["tags"], result["_meta"]["tags"] = get_tags_with_meta(
+                        self.nodestore_event
+                    )
+            result["children"] = [child.to_dict(full, detailed) for child in self.children]
+        return result
+
 
-def find_event(items, function, default=None):
+# Can't do Callable[[Any], bool] for the function cause anything on an Any -> Any
```

Comment:
Can you elaborate on this?

Reply:
Hmm actually looking again at this, I think my comment is wrong and I can type this
```python
def find_event(items: Sequence[Mapping[str, str]], function: Callable[[Mapping[str, str]], bool], default: Optional[Mapping[str, str]] = None) -> Optional[Mapping[str, str]]:
```

But now its complaining about https://github.com/getsentry/sentry/pull/25432/files#diff-619b86f549a9f11a566ca757be8da43a63bcaf6e4d7d3fe88a4222937a0209c6R274 being not indexed 🤔 I'll keep trying